dsd/constraints.py

```diff
@@ -616,17 +644,48 @@ def generate_sequence(self, rng: np.random.Generator) -> str:
 
         :param rng:
             numpy random number generator to use. To use a default, pass :py:data:`np.default_rng`.
+        :param previous_sequence:
+            previously generated sequence to be replaced by a new sequence; None if no previous
+            sequence exists. Used in :py:meth:`DomainPool.find_steps_distance_sequences`
+            to choose a new sequence "close" to itself in Hamming distance. The number of
+            differences between previous_sequence and its neighbors is determined by randomly
+            picking a Hamming distance from :py:data:`DomainPool.hamming_probability` with
+            weighted probabilities of choosing each distance.
         :return:
             DNA sequence of given length satisfying :py:data:`DomainPool.numpy_constraints` and
             :py:data:`DomainPool.sequence_constraints`
         """
         log_debug_sequence_constraints_accepted = False
-        sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
-        while not self.satisfies_sequence_constraints(sequence):
-            logger.debug(f'rejecting domain sequence {sequence}; failed some sequence constraint')
+        if not self.replace_with_close_sequences or not previous_sequence:
+            # takes a completely random sequence from domain pool
             sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
-        if log_debug_sequence_constraints_accepted:
-            logger.debug(f'accepting domain sequence {sequence}; passed all sequence constraints')
+            while not self.satisfies_sequence_constraints(sequence):
+                logger.debug(f'rejecting domain sequence {sequence}; failed some sequence constraint')
+                sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
+            if log_debug_sequence_constraints_accepted:
+                logger.debug(f'accepting domain sequence {sequence}; passed all sequence constraints')
+        else:
+            # takes neighbor to previous sequence; difference in bases randomly chosen
+            hamming_distances = []
+            hamming_probabilities = []
+            for key, val in self.hamming_probability.items(): # make lists for np.random.choice
```

Comment:
This loop can be more succinctly replaced by 

```python
hamming_distances = list(self.hamming_probability.keys())
hamming_probabilities = list(self.hamming_probability.values())
```

See https://stackoverflow.com/questions/21869973/turn-a-dictionary-into-two-lists-one-for-keys-and-the-other-for-values/21870002